pkg/plugins/backendplugin/instrumentation.go

```diff
@@ -0,0 +1,45 @@
+package backendplugin
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	pluginRequestCounter  *prometheus.CounterVec
+	pluginRequestDuration *prometheus.SummaryVec
+)
+
+func init() {
+	pluginRequestCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "plugin_request_total",
+		Help: "The total amount of plugin requests",
+	}, []string{"plugin_id", "endpoint", "status"})
+
+	pluginRequestDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Name:       "plugin_request_duration_milliseconds",
+		Help:       "Plugin request duration",
+		Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+	}, []string{"plugin_id", "endpoint"})
```

Comment:
Nit. After further consideration think we should prefix these metrics with `grafana_` by our convention. What do you think?

Almost all of grafana's custom metrics are prefixed with `grafana_`:
```txt
# HELP grafana_alerting_active_alerts amount of active alerts
# TYPE grafana_alerting_active_alerts gauge
grafana_alerting_active_alerts 33
# HELP grafana_alerting_execution_time_milliseconds summary of alert exeuction duration
# TYPE grafana_alerting_execution_time_milliseconds summary
grafana_alerting_execution_time_milliseconds{quantile="0.5"} 0
grafana_alerting_execution_time_milliseconds{quantile="0.9"} 1
grafana_alerting_execution_time_milliseconds{quantile="0.99"} 1
grafana_alerting_execution_time_milliseconds_sum 5
grafana_alerting_execution_time_milliseconds_count 17
# HELP grafana_alerting_result_total alert execution result counter
# TYPE grafana_alerting_result_total counter
grafana_alerting_result_total{state="alerting"} 8
grafana_alerting_result_total{state="ok"} 9
# HELP grafana_api_admin_user_created_total api admin user created counter
# TYPE grafana_api_admin_user_created_total counter
grafana_api_admin_user_created_total 0
# HELP grafana_api_dashboard_get_milliseconds summary for dashboard get duration
# TYPE grafana_api_dashboard_get_milliseconds summary
grafana_api_dashboard_get_milliseconds{quantile="0.5"} NaN
grafana_api_dashboard_get_milliseconds{quantile="0.9"} NaN
grafana_api_dashboard_get_milliseconds{quantile="0.99"} NaN
grafana_api_dashboard_get_milliseconds_sum 0
grafana_api_dashboard_get_milliseconds_count 0
# HELP grafana_api_dashboard_save_milliseconds summary for dashboard save duration
# TYPE grafana_api_dashboard_save_milliseconds summary
grafana_api_dashboard_save_milliseconds{quantile="0.5"} NaN
grafana_api_dashboard_save_milliseconds{quantile="0.9"} NaN
grafana_api_dashboard_save_milliseconds{quantile="0.99"} NaN
grafana_api_dashboard_save_milliseconds_sum 0
grafana_api_dashboard_save_milliseconds_count 0
# HELP grafana_api_dashboard_search_milliseconds summary for dashboard search duration
# TYPE grafana_api_dashboard_search_milliseconds summary
grafana_api_dashboard_search_milliseconds{quantile="0.5"} NaN
grafana_api_dashboard_search_milliseconds{quantile="0.9"} NaN
grafana_api_dashboard_search_milliseconds{quantile="0.99"} NaN
grafana_api_dashboard_search_milliseconds_sum 0
grafana_api_dashboard_search_milliseconds_count 0
# HELP grafana_api_dashboard_snapshot_create_total dashboard snapshots created
# TYPE grafana_api_dashboard_snapshot_create_total counter
grafana_api_dashboard_snapshot_create_total 0
# HELP grafana_api_dashboard_snapshot_external_total external dashboard snapshots created
# TYPE grafana_api_dashboard_snapshot_external_total counter
grafana_api_dashboard_snapshot_external_total 0
# HELP grafana_api_dashboard_snapshot_get_total loaded dashboards
# TYPE grafana_api_dashboard_snapshot_get_total counter
grafana_api_dashboard_snapshot_get_total 0
# HELP grafana_api_dataproxy_request_all_milliseconds summary for dataproxy request duration
# TYPE grafana_api_dataproxy_request_all_milliseconds summary
grafana_api_dataproxy_request_all_milliseconds{quantile="0.5"} NaN
grafana_api_dataproxy_request_all_milliseconds{quantile="0.9"} NaN
grafana_api_dataproxy_request_all_milliseconds{quantile="0.99"} NaN
grafana_api_dataproxy_request_all_milliseconds_sum 0
grafana_api_dataproxy_request_all_milliseconds_count 0
# HELP grafana_api_login_oauth_total api login oauth counter
# TYPE grafana_api_login_oauth_total counter
grafana_api_login_oauth_total 0
# HELP grafana_api_login_post_total api login post counter
# TYPE grafana_api_login_post_total counter
grafana_api_login_post_total 0
# HELP grafana_api_login_saml_total api login saml counter
# TYPE grafana_api_login_saml_total counter
grafana_api_login_saml_total 0
# HELP grafana_api_models_dashboard_insert_total dashboards inserted 
# TYPE grafana_api_models_dashboard_insert_total counter
grafana_api_models_dashboard_insert_total 0
# HELP grafana_api_org_create_total api org created counter
# TYPE grafana_api_org_create_total counter
grafana_api_org_create_total 0
# HELP grafana_api_response_status_total api http response status
# TYPE grafana_api_response_status_total counter
grafana_api_response_status_total{code="200"} 7
grafana_api_response_status_total{code="404"} 0
grafana_api_response_status_total{code="500"} 0
grafana_api_response_status_total{code="unknown"} 0
# HELP grafana_api_user_signup_completed_total amount of users who completed the signup flow
# TYPE grafana_api_user_signup_completed_total counter
grafana_api_user_signup_completed_total 0
# HELP grafana_api_user_signup_invite_total amount of users who have been invited
# TYPE grafana_api_user_signup_invite_total counter
grafana_api_user_signup_invite_total 0
# HELP grafana_api_user_signup_started_total amount of users who started the signup flow
# TYPE grafana_api_user_signup_started_total counter
grafana_api_user_signup_started_total 0
# HELP grafana_aws_cloudwatch_get_metric_data_total counter for getting metric data time series from aws
# TYPE grafana_aws_cloudwatch_get_metric_data_total counter
grafana_aws_cloudwatch_get_metric_data_total 0
# HELP grafana_aws_cloudwatch_get_metric_statistics_total counter for getting metric statistics from aws
# TYPE grafana_aws_cloudwatch_get_metric_statistics_total counter
grafana_aws_cloudwatch_get_metric_statistics_total 0
# HELP grafana_aws_cloudwatch_list_metrics_total counter for getting list of metrics from aws
# TYPE grafana_aws_cloudwatch_list_metrics_total counter
grafana_aws_cloudwatch_list_metrics_total 0
# HELP grafana_build_info A metric with a constant '1' value labeled by version, revision, branch, and goversion from which Grafana was built
# TYPE grafana_build_info gauge
grafana_build_info{branch="plugin_metrics",edition="oss",goversion="go1.14.1",revision="9acfe1f",version="7.0.0-pre"} 1
# HELP grafana_db_datasource_query_by_id_total counter for getting datasource by id
# TYPE grafana_db_datasource_query_by_id_total counter
grafana_db_datasource_query_by_id_total 17
# HELP grafana_instance_start_total counter for started instances
# TYPE grafana_instance_start_total counter
grafana_instance_start_total 1
# HELP grafana_ldap_users_sync_execution_time summary for LDAP users sync execution duration
# TYPE grafana_ldap_users_sync_execution_time summary
grafana_ldap_users_sync_execution_time{quantile="0.5"} NaN
grafana_ldap_users_sync_execution_time{quantile="0.9"} NaN
grafana_ldap_users_sync_execution_time{quantile="0.99"} NaN
grafana_ldap_users_sync_execution_time_sum 0
grafana_ldap_users_sync_execution_time_count 0
# HELP grafana_page_response_status_total page http response status
# TYPE grafana_page_response_status_total counter
grafana_page_response_status_total{code="200"} 16
grafana_page_response_status_total{code="404"} 0
grafana_page_response_status_total{code="500"} 0
grafana_page_response_status_total{code="unknown"} 0
# HELP grafana_plugin_build_info A metric with a constant '1' value labeled by pluginId, pluginType and version from which Grafana plugin was built
# TYPE grafana_plugin_build_info gauge
grafana_plugin_build_info{plugin_id="agenty-flowcharting-panel",plugin_type="panel",version="0.6.0"} 1
grafana_plugin_build_info{plugin_id="google-sheets-datasource",plugin_type="datasource",version="0.9.0-dev"} 1
grafana_plugin_build_info{plugin_id="grafana-image-renderer",plugin_type="renderer",version="1.0.12"} 1
grafana_plugin_build_info{plugin_id="grafana-piechart-panel",plugin_type="panel",version="1.3.9"} 1
grafana_plugin_build_info{plugin_id="grafana-polystat-panel",plugin_type="panel",version="1.0.16"} 1
grafana_plugin_build_info{plugin_id="grafana-test-datasource",plugin_type="datasource",version="0.1.0"} 1
grafana_plugin_build_info{plugin_id="grafana-worldmap-panel",plugin_type="panel",version="0.2.0"} 1
grafana_plugin_build_info{plugin_id="marefr-covid-19-datasource",plugin_type="datasource",version="0.1.0"} 1
grafana_plugin_build_info{plugin_id="marefr-marefr-vega-panel",plugin_type="panel",version="1.0.0"} 1
grafana_plugin_build_info{plugin_id="raintank-worldping-app",plugin_type="app",version="1.2.6"} 1
grafana_plugin_build_info{plugin_id="worldping-cta",plugin_type="panel",version=""} 1
grafana_plugin_build_info{plugin_id="worldping-endpoint-list",plugin_type="panel",version=""} 1
grafana_plugin_build_info{plugin_id="worldping-endpoint-nav",plugin_type="panel",version=""} 1
# HELP grafana_proxy_response_status_total proxy http response status
# TYPE grafana_proxy_response_status_total counter
grafana_proxy_response_status_total{code="200"} 0
grafana_proxy_response_status_total{code="404"} 0
grafana_proxy_response_status_total{code="500"} 0
grafana_proxy_response_status_total{code="unknown"} 0
# HELP grafana_stat_active_users number of active users
# TYPE grafana_stat_active_users gauge
grafana_stat_active_users 3
# HELP grafana_stat_total_orgs total amount of orgs
# TYPE grafana_stat_total_orgs gauge
grafana_stat_total_orgs 19
# HELP grafana_stat_total_playlists total amount of playlists
# TYPE grafana_stat_total_playlists gauge
grafana_stat_total_playlists 2
# HELP grafana_stat_total_users total amount of users
# TYPE grafana_stat_total_users gauge
grafana_stat_total_users 21
# HELP grafana_stat_totals_active_admins total amount of active admins
# TYPE grafana_stat_totals_active_admins gauge
grafana_stat_totals_active_admins 3
# HELP grafana_stat_totals_active_editors total amount of active editors
# TYPE grafana_stat_totals_active_editors gauge
grafana_stat_totals_active_editors 0
# HELP grafana_stat_totals_active_viewers total amount of viewers
# TYPE grafana_stat_totals_active_viewers gauge
grafana_stat_totals_active_viewers 0
# HELP grafana_stat_totals_admins total amount of admins
# TYPE grafana_stat_totals_admins gauge
grafana_stat_totals_admins 5
# HELP grafana_stat_totals_dashboard total amount of dashboards
# TYPE grafana_stat_totals_dashboard gauge
grafana_stat_totals_dashboard 727
# HELP grafana_stat_totals_editors total amount of editors
# TYPE grafana_stat_totals_editors gauge
grafana_stat_totals_editors 5
# HELP grafana_stat_totals_viewers total amount of viewers
# TYPE grafana_stat_totals_viewers gauge
grafana_stat_totals_viewers 10
```

Reply:
Good catch. I exposed the exporter name in metrics.go and use that as namespace.